src/lib.rs

```diff
@@ -89,6 +99,46 @@ impl Deepgram {
     ///
     /// Panics under the same conditions as [`reqwest::Client::new`].
     pub fn new<K: AsRef<str>>(api_key: K) -> Self {
+        Self::with_base_url(api_key, DEEPGRAM_BASE_URL)
+    }
+
+    /// Construct a new Deepgram client with the specified base URL.
+    ///
+    /// When using a self-hosted instance of deepgram, this will be the
+    /// host portion of your own instance. For instance, if you would
+    /// query your deepgram instance at `http://deepgram.internal/v1/listen`,
+    /// the base_url will be `http://deepgram.internal`.
+    ///
+    /// Admin features, such as billing, usage, and key management will
+    /// still go through the hosted site at `https://api.deepgram.com`.
+    ///
+    /// Create your first API key on the [Deepgram Console][console].
+    ///
+    /// [console]: https://console.deepgram.com/
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// # use deepgram::Deepgram;
+    /// let deepgram = Deepgram::with_base_url(
+    ///     "apikey12345",
+    ///     "http://localhost:8080",
+    /// );
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics under the same conditions as [`reqwest::Client::new`], or if `base_url`
+    /// is not a valid URL.
+    ///
+    ///
+
+    pub fn with_base_url<K, U>(api_key: K, base_url: U) -> Self
```

Comment:
I find the signature of this function confusing for use with self-hosted. It seems to imply that self-hosted requests require an API key, when that is not the case. Even more concerning, if users did provide an API key, we would unnecessarily be passing that API key onto the self-hosted instance. We probably don't want to be doing that.

Reply:
That makes sense.  I'll make it optional here, and explain the asymmetry between the signatures of `with_base_url` and `new` in the docs.

Reply:
So having an `Option<impl AsRef<str>>` for `with_base_url` was confusing, because it meant that omitting it would require you to write `None::<String>` (or some other type that matches the generic constraint.  Instead I'm making two functions. `with_base_url` and `with_base_url_and_api_key`.  I'm keeping the api_key-only method named `new` since most users will just use that interface.  I'm almost at the point of wanting a builder pattern here.